fendermint/vm/interpreter/src/fvm/state/snapshot.rs

```diff
@@ -0,0 +1,357 @@
+use anyhow::anyhow;
+use cid::multihash::{Code, MultihashDigest};
+use cid::Cid;
+use futures_core::Stream;
+use fvm::state_tree::StateTree;
+use fvm_ipld_blockstore::Blockstore;
+use fvm_ipld_car::{load_car_unchecked, CarHeader};
+use fvm_ipld_encoding::{from_slice, DAG_CBOR};
+use libipld::Ipld;
+use std::collections::VecDeque;
+use std::path::Path;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tokio_stream::StreamExt;
+use tokio_util::compat::{TokioAsyncReadCompatExt, TokioAsyncWriteCompatExt};
+
+pub type BlockHeight = u64;
+
+/// Taking snapshot of the current blockchain state
+pub enum Snapshot<DB> {
+    V1(V1Snapshot<DB>),
+}
+
+impl<DB> Snapshot<DB>
+where
+    DB: Blockstore + Clone + 'static + Send,
+{
+    pub fn new(
+        store: DB,
+        state_params: FvmStateParams,
+        block_height: BlockHeight,
+    ) -> anyhow::Result<Self> {
+        Ok(Self::V1(V1Snapshot::new(
+            store,
+            state_params,
+            block_height,
+        )?))
+    }
+
+    /// Read the snapshot from file and load all the data into the store
+    pub async fn read_car(path: impl AsRef<Path>, store: DB) -> anyhow::Result<Self> {
+        let file = tokio::fs::File::open(path).await?;
+        let snapshot = V1Snapshot::read_car(file, store).await?;
+        Ok(Self::V1(snapshot))
+    }
+
+    /// Write the snapshot to car file
+    pub async fn write_car(self, path: impl AsRef<Path>) -> anyhow::Result<()> {
+        let file = tokio::fs::File::create(path).await?;
+
+        // derive the car header roots
+        let car = CarHeader::new(self.car_header_roots()?, self.version());
+
+        match self {
+            Snapshot::V1(s) => s.write_car(car, file).await?,
+        }
+
+        Ok(())
+    }
+
+    pub fn version(&self) -> u64 {
+        match self {
+            Snapshot::V1(_) => 1,
+        }
+    }
+
+    /// Init the block chain from the snapshot
+    pub fn init_chain(&self) -> anyhow::Result<()> {
+        todo!()
+    }
+
+    fn car_header_roots(&self) -> anyhow::Result<Vec<Cid>> {
+        match self {
+            Snapshot::V1(s) => s.car_header_roots(),
+        }
+    }
+}
+
+pub struct V1Snapshot<DB> {
+    /// The root cid of the state tree
+    state_tree_root: Cid,
+    /// The state tree of the current blockchain
+    state_tree: StateTree<ReadOnlyBlockstore<DB>>,
+    /// The latest block height with state params serialized to bytes
+    block_state_params_bytes: Vec<u8>,
+    /// Block state params cid derived from block state params bytes
+    block_state_params_cid: Cid,
+}
+
+impl<DB> V1Snapshot<DB>
+where
+    DB: Blockstore + Clone + 'static + Send,
+{
+    /// Creates a new V2Snapshot struct. Caller ensure store
+    pub fn new(
+        store: DB,
+        state_params: FvmStateParams,
+        block_height: BlockHeight,
+    ) -> anyhow::Result<Self> {
+        let state_tree =
+            StateTree::new_from_root(ReadOnlyBlockstore::new(store), &state_params.state_root)?;
```

Comment:
Because `FvmStateParams` have a `state_root` inside it, and it's a `Cid`, isn't it enough to use the CID of the `(state_params, block_height)` structure as a single root CID for the CAR file? The state root is implicitly reachable through that.

Reply:
That works too